### ISSUE TITLE
Parallel queries

### ISSUE THREAD

**Issue by pavelvondrasek** [linq2db/linq2db#2495]
Hi, 

I believe you are safe! I have question.

I thought about awaiting queries. I tried parallel queries like this:
```cs
var userTask = _userRepository.GetUser();
var localesTask = _contentRepository.GetLocales();
var contentsTask = _contentRepository.GetContents();

await Task.WhenAll(userTask, localesTask, contentsTask);
```

I read about `DataConnection` and `DataContext` but I'm bit confused.

Do you recommend me use parallel queries for better performance or not?
And how you reuse database connections?

Now I'm resuing connection in my repository and dispose after repository is destructed.
I tried to create new `DataConnection` for every query and I think my api is slower. 

I would like to know what is best practice for async and parallel queries. :)

Thank you very much and stay safe! :)

**Comment by sdanyliv** [comment]
`DataConnection` and `DataContext` are not thread safe. So you have to create new instances of `DataConnection` for each parallel task.
If you reusing DataConnection in repositories only this variant is acceptable:
```cs
var userTask = await _userRepository.GetUser();
var localesTask = await contentRepository.GetLocales();
var contentsTask = await _contentRepository.GetContents();
```

**Comment by pavelvondrasek** [comment]
Ok, and if will create new instance for each parallel task can I eleminate database reconnection?

I tested that with paralllel tasks and new `DataConnection` is my API response 2x slower (maybe it's slow internet connection with high ping at my home).

Is connection pooling solution: https://www.npgsql.org/doc/connection-string-parameters.html#pooling ? :)

**Comment by sdanyliv** [comment]
So specify pooling parameters in Connection String. `DataConnection` is very lightweight class and its creation do almost nothing, but opening new connection can be slow and pooling may help.

**Comment by viceroypenguin** [comment]
Slight correction. With the `await`s in-line, those queries aren't actually running in parallel. This variant will, as long as each repository creates it's own `DataConnection`:

````
var userTask = _userRepository.GetUser();
var localesTask = __contentRepository.GetLocales();
var contentsTask = _contentRepository.GetContents();

var users = await userTask;
var locales = await localesTask;
var contents = await contentsTask;
```

I use this model in my code regularly.

**Comment by sdanyliv** [comment]
Closing for now, will reopen if needed.

**Comment by dipique** [comment]
For those who find this page in search results looking for how to run linq2db queries in parallel, I've been able to use this pattern successfully:

```
IEnumerable<int> recordIds = new [] { 4, 13, 44, 92 }; // start with a collection that contains one item per desired parallel query
var results = recordIds
    .AsParallel() // run the rest of the linq evaluation in parallel rather than sequentially
    .Select(recordId => {
        // beginning of parallel query
        using (var db = new MyDB()) // use a new data context for every parallel query
        {                           // or create a pool if you feel ambitious
            // run your query, retrieving any data you want back
            var deletedCount = db.MyTable
                .Where(record => record.Id == recordId)
                .Delete();
                
            // return whatever results you want from this query
            return (recordId, deletedCount);
        }
    });

// PLINQ (Parallel Linq) uses deferred execution, so no queries have
// been run yet. Any typical method of running it will work such as
// .ToArray() or .ToList(). The .Sum() method below also works. This
// will be intuitive to anyone used to LINQ or Linq2db.

// example of using the results
var totalRecordsDeleted = results.Sum(result => result.deletedCount);
```

Be careful with methods like `.Any()`, `.First()`, and so on that will not necessarily force FULL execution. The behavior for partially-executed PLINQ partitions is complex and better avoided if possible. As long as your execution/materialization method forces iteration through the results of every query, you won't have to deal with that.

This pattern doesn't require `await/async` which some may consider an advantage (or at least situationally useful).

**Comment by sdanyliv** [comment]
Well, probably it is bad example. This query will beat your "Parallel execution" by resources and performance:
```cs
IEnumerable<int> recordIds = new [] { 4, 13, 44, 92 }; 

var deletedCount = db.MyTable
    .Where(record => recordIds.Contains(record.Id)
    .Delete();
```

**Comment by dipique** [comment]
> Well, probably it is bad example. This query will beat your "Parallel execution" by resources and performance:
> 
> ```cs
> IEnumerable<int> recordIds = new [] { 4, 13, 44, 92 }; 
> 
> var deletedCount = db.MyTable
>     .Where(record => recordIds.Contains(record.Id)
>     .Delete();
> ```

My purpose was to demonstrate running parallel queries, not to identify which queries ought to be run in parallel.

**Comment by ademchenko** [comment]
`DataConnection and DataContext are not thread-safe. So you have to create new instances of DataConnection for each parallel task.`
@sdanyliv , is it just an implementation detail of linq2db or the underlying mechanisms also cannot proceed that way? I.e. suppose you have made the proper fixes of DataConnection and DataContext. Will then be possible to post parallel queries into a single connection having that it is just a wrapper over ADO Connection?

We have just also faced an issue with the thread safety of DataContext, and I'm thinking if that's a perspective way to fix these bugs or, anyway, the other issues on a lower level will arise.

**Comment by MaceWindu** [comment]
this is underlying issue. ado.net is not thread-safe
